### PR TITLE
Pp order register

### PR DESCRIPTION
**Pull request by vmchale**
This is a first cut for a fix of https://github.com/haskell/cabal/issues/55 aka https://github.com/haskell/cabal/issues/1906. As discussed in a [previous PR](https://github.com/haskell/cabal/pull/6233), the "proper" way to fix c2hs handling (in a way that can be generalized) is to present the modules in question to the preprocessor, which can request a particular ordering (using its own lexer, etc.)

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

This makes modifications in the same places as https://github.com/haskell/cabal/pull/6233 so I am confident it works.

### CODE REVIEW

**Comment by emilypi** [comment]
@phadej  do you mind taking a look at this?

File: Cabal/src/Distribution/Simple/PreProcess.hs [hunk 1]
```diff
@@ -202,6 +227,8 @@ preprocessComponent pd comp lbi clbi isSrcDist verbosity handlers = do
           die' verbosity $ "No support for preprocessing benchmark "
                         ++ "type " ++ prettyShow tt
   where
+    orderingFromHandlers v d = threadM . fmap (($d) . ($v) . ppOrdering . snd)
+    threadM = foldr (>=>) pure
```
Comment (emilypi) [inline Cabal/src/Distribution/Simple/PreProcess.hs:231]: It might be simpler to just do

```haskell
orderingFromHandlers v d hndlrs mods =
  foldM (\acc (_,pp) -> ppOrdering pp v d acc) mods hndlrs
```

**Review by emilypi — COMMENTED** [review]
@vmchale it would be good to add test coverage for this

File: Cabal/src/Distribution/Simple/PreProcess.hs (hunk 1 above)
Comment (vmchale) [reply Cabal/src/Distribution/Simple/PreProcess.hs]: Sure. That is less indirect!

**Comment by vmchale** [comment]
I've spot tested this with the local exchange, with the following results:

```
vanessa@system76-pc:~/dev/bitnomial$ curl 'localhost:11030/exchange/orders/customer?clearing_account_id=1'
[[1,[[13,"b17301a2a"],[14,"b17301a2a"]]]]vanessa@system76-pc:~/dev/bitnomial$
```

(made orders myself in the gateway client; they showed up!)

**Comment by Mikolaj** [comment]
Something, probably shallow, is borked with the tests. Let me try to update wrt master branch and rerun the tests, just in case the breakage comes from master.

**Comment by Mikolaj** [comment]
Borked still, but let me try again.

**Comment by Mikolaj** [comment]
@vmchale: well, the CI failure is not shallow after all: `Not in scope: ppOrdering` in the testsuite. Would you find time to fix?

**Comment by Mikolaj** [comment]
So, I guess, we are fine regarding CI? @vmchale, would you like to rebase, which gets rid of all the merge commits, or should we squash or not create a commit mentioning the PR and delimiting all others at all?

**Review by Mikolaj — COMMENTED** [review]
LGTM, but what's with all those deleted changelog files? Is that some merge artefact? Would a rebase sort this out perhaps?

**Comment by vmchale** [comment]
@Mikolaj I've rebased + force-pushed! It can be squashed to merge if you like.

**Review by Mikolaj — APPROVED** [review]

**Comment by Mikolaj** [comment]
@vmchale: looks fine now; I think we mya just merge with a PR commit as soon as @emilypi or @fgaz review.

**Review by emilypi — APPROVED** [review]
LGTM - thanks @vmchale :smile:
